src/FileFormats/format_flatbuffer_opl3.cpp

```diff
@@ -0,0 +1,206 @@
+                    break;
+            }
+
+            ins->fine_tune = instrument->secondVoiceTuning();
+            int mode = instrument->mode();
+            ins->en_pseudo4op = (mode == Mode_Pseudo);
+            ins->en_4op = (mode == Mode_Pseudo) || (mode == Mode_FourOp);
+            ins->percNoteNum = instrument->percussionKey();
+            ins->note_offset1 = instrument->keyOffset1();
+            ins->note_offset2 = instrument->keyOffset2();
+            ins->velocity_offset = instrument->velocityOffset();
+            ins->is_blank = instrument->blank();
+            ins->ms_sound_kon = instrument->konMs();
+            ins->ms_sound_koff = instrument->koffMs();
+            strncpy(ins->name, instrument->name()->c_str(), 32);
+
+            ins->setFBConn1(instrument->fb_conn1());
+            ins->setFBConn2(instrument->fb_conn2());
+
+            ins->setAVEKM(MODULATOR1,    instrument->modulator1()->AVEKM());
+            ins->setAtDec(MODULATOR1,    instrument->modulator1()->AtDec());
+            ins->setSusRel(MODULATOR1,   instrument->modulator1()->SusRel());
+            ins->setWaveForm(MODULATOR1, instrument->modulator1()->WaveForm());
+            ins->setKSLL(MODULATOR1,     instrument->modulator1()->KSLL());
+
+            ins->setAVEKM(CARRIER1,    instrument->carrier1()->AVEKM());
+            ins->setAtDec(CARRIER1,    instrument->carrier1()->AtDec());
+            ins->setSusRel(CARRIER1,   instrument->carrier1()->SusRel());
+            ins->setWaveForm(CARRIER1, instrument->carrier1()->WaveForm());
+            ins->setKSLL(CARRIER1,     instrument->carrier1()->KSLL());
+
+            ins->setAVEKM(MODULATOR2,    instrument->modulator2()->AVEKM());
+            ins->setAtDec(MODULATOR2,    instrument->modulator2()->AtDec());
+            ins->setSusRel(MODULATOR2,   instrument->modulator2()->SusRel());
+            ins->setWaveForm(MODULATOR2, instrument->modulator2()->WaveForm());
+            ins->setKSLL(MODULATOR2,     instrument->modulator2()->KSLL());
+
+            ins->setAVEKM(CARRIER2,    instrument->carrier2()->AVEKM());
+            ins->setAtDec(CARRIER2,    instrument->carrier2()->AtDec());
+            ins->setSusRel(CARRIER2,   instrument->carrier2()->SusRel());
+            ins->setWaveForm(CARRIER2, instrument->carrier2()->WaveForm());
+            ins->setKSLL(CARRIER2,     instrument->carrier2()->KSLL());
+        }
+
+        switch (bankType) {
+            case BankType_Melodic:
+                current_melodic_bank++;
+                break;
+            case BankType_Percussion:
+                current_percusive_bank++;
+                break;
+        }
+    }
+
+    return FfmtErrCode::ERR_OK;
+}
+
+FfmtErrCode FlatbufferOpl3::saveFile(QString filePath, FmBank &bank)
+{
+    QFile file(filePath);
+    if(!file.open(QIODevice::WriteOnly))
+        return FfmtErrCode::ERR_NOFILE;
+
+    return FfmtErrCode::ERR_OK;
```

Comment:
I think, you better to put "ERR_NOT_IMPLEMENTED" error until you will make the support of file saving. Also, when you making new formats, you better implement SAVE before to implement OPEN as before current moment, you have no sample files to test them. Implement OPEN first when you gonna support already existing file formats and you have some sample files to use them while polishing the parser or generator.

Reply:
yes obviously... it was just a quick edit.
I'm going to implement the save ASAP